8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

pabulaner opened the pull request:
This pull request fixes the system menu bar on MacOS when combining windows of Swing and JavaFX.

# Behavior before

If for some reason You needed to initialize AWT before JavaFX and You wanted to install the system menu bar from the JavaFX side, this wasn't possible. This issue is persistent even when You didn't open a Swing window. 

One scenario where this kind of issue happens is when You use install4j (see https://www.ej-technologies.com/install4j). In this case AWT is initialized by install4j and therefore You can't use the JavaFX system menu bar anymore.


# Behavior after

The fix allows JavaFX to install a system menu bar even if it is initialized after AWT. This is achieved by only changing code inside JavaFX. Each JavaFX window stores the previously installed menu bar when gaining focus and will restore this menu bar if the focus was lost. This only happens if the system menu bar installed by the JavaFX window is still unchanged.


# Tests

This PR introduces tests for the system menu bar in addition to verifying its own behavior / changes. The tests include single- and multi-window tests while interacting with Swing. The tests ensure that the menu bar stays the same for each window, no matter how You switch focus between them.


# Additional benifits 

This fix is not specifically for AWT, but allows JavaFX to interact much more compatibly with other frameworks that make use of the system menu bar.


# Review from AWT

In the previous PR related to this one, the comment was made that the folks from AWT should take a look at this fix. It would be great and much appreciated if someone could initiate it.


# Add disable flag?

We could also add a flag to prevent JavaFX from installing a system menu bar for users who have found other fixes for their projects / setups. This could be used to restore the previous behavior when AWT is initialized first.


Co-Author: @FlorianKirmaier

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8359108](https://bugs.openjdk.org/browse/JDK-8359108): Mac - When Swing starts First, native application menu doesn't work for JavaFX (**Bug** - P3)


### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Martin Fox](https://openjdk.org/census#mfox) (@beldenfox - Committer)

### Contributors
 * Florian Kirmaier `<fkirmaier@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1904/head:pull/1904` \
`$ git checkout pull/1904`

Update a local copy of the PR: \
`$ git checkout pull/1904` \
`$ git pull https://git.openjdk.org/jfx.git pull/1904/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1904`

View PR using the GUI difftool: \
`$ git pr show -t 1904`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1904.diff">https://git.openjdk.org/jfx/pull/1904.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1904#issuecomment-3292992538)
</details>

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

# Conflicts:
#	modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Overrides.m

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth commented:
This approach seems much better than the previous in that it is activated only when a JavaFX Stage becomes the focused (key) window, and it saves / restores the menu such that whenever an different (e.g., AWT) window is focused, the AWT menu will be installed.

In addition to reviewing and checking that it only touches the menu when it should, testing will be critical.

Reviewers: @kevinrushforth @andy-goryachev-oracle @beldenfox
Cc: @prrace @honkar-jdk

/reviewers 3

kevinrushforth commented:
As mentioned above, testing will be key here. This needs to work reliably in the case of a multi-window app when the application is hidden and restored, and when switching between the windows of the app. The automated tests should already cover some of the cases. There may be others to consider.

Here are a few cases that I can think of to check:

* Pure FX app: 2 FX Stages (AWT toolkit not initialized at all) -- the fix should do nothing
* 2 FX Stages with FX initialized first (AWT is initialized last, but no AWT/Swing windows) -- the fix should do nothing
* 2 FX Stages with AWT initialized first (but no AWT/Swing windows) -- the fix should install the FX window when either Stage is focused (there will not be any AWT menu)
* a mix of AWT Windows and FX Stages with FX initialized first -- FX will install the menu and keep it there, since AWT won't install a menu when it doesn't own the NSApplication
* a mix of AWT Windows and FX Stages with FX initialized first -- AWT menu will be installed when an AWT window is active, FX menu will be installed when an FX Stage is active
* 2 AWT Stages one of which has a JFXPanel with an FX scene, initialize AWT first -- fix should be a noop
* 2 AWT Stages one of which has a JFXPanel with an FX scene, initialize FX first -- fix should be a noop

I presume that the automated tests will cover some of these, but we might want additional tests.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

pabulaner commented:
I have now also added the JFXPanel test, but only for the first of the two tests involving the JFXPanel (so where AWT is initialized first). In the other case AWT won't install a menu bar, since JavaFX was initialized first, so nothing will happen.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth reviewed (commented):
This is a very preliminary review. I haven't run it yet, but I will soon. I left a few inline comments of things I happened to spot.

The fix itself looks good to me. 

The test framework and set of tests looks fine on first glance, but I haven't looked at them in detail. I will do so, along with running them

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth commented on tests/system/src/test/java/test/com/sun/glass/ui/mac/MacOSSystemMenuTestBase.java:
```diff
@@ -0,0 +1,444 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class MacOSSystemMenuTestBase {
```
Somewhere in this base class, add a static method annotated with `@BeforeAll` that does an `assumeTrue(PlatformUtil.isMac())` to ensure that these tests don't execute on the other platforms.

kevinrushforth reviewed (commented):
I ran the tests today on my Mac M1 system. I left one comment inline about needing to ensure that these tests only run on macOS.

The following two tests fail consistently for me:

```
MacOSSystemMenuMultiWindowFXOnlySwingLast
MacOSSystemMenuMultiWindowTest
```

The following fail intermittently:

```
MacOSSystemMenuSetWithSwingTestFirst
MacOSSystemMenuJFXPanelSwingFirstTest
```

Also, I noticed that running all 7 tests takes over 6 minutes on my system. I tracked down the reason to the large number of calls to fork the `osascript` process.

Here is the count by test:


Test name | Number of calls
--------- | ---------------
MacOSSystemMenuJFXPanelSwingFirstTest | 253 
MacOSSystemMenuMultiWindowFXOnlySwingFirst | 241 
MacOSSystemMenuMultiWindowFXOnlySwingLast | 276 
MacOSSystemMenuMultiWindowTest | 276 
MacOSSystemMenuMultiWindowWithSwingFirstTest | 407 
MacOSSystemMenuSetWithSwingTestFirst | 466 
MacOSSystemMenuSingleWindowWithSwingFirstTest | 165 

It isn't a deal-breaker, but it would be nice to cut this down a bit if without sacrificing test coverage (e.g,. I think it's good to test with at least one sub-menu and menus with different number of items)

kevinrushforth commented:
I note that your branch is very out of date. Please merge the latest master into your branch. There is a minor merge conflict in `modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Overrides.m` that you will need to resolve. Leave the commented out `NSLog` (which is present in the latest master and conflicts with your change) as the first line of the function.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth commented:
> I have rebased to the master

To echo what the Skara said: next time, please _merge_ master instead of rebasing to master.. Rebasing makes reviewing new change since last review difficult at best.

> for me everything seems to work. Could You maybe provide me with some more information what exactly doesn't work with the tests for You?

Sure. I'll provide the logs.

kevinrushforth commented on modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Overrides.m:
```diff
@@ -69,6 +69,10 @@ - (void)windowDidChangeScreen:(NSNotification *)notification
+    // just store a default menu
+    self->hostMenu = [NSApp mainMenu];
+
     //NSLog(@"windowDidBecomeKey: %p", self);
```
Minor merge issue: This should be the first line of this method.

kevinrushforth reviewed (commented):
I left a couple comments about a (minor) merge issue.

pabulaner commented:
As long as manual tests are considered a good practice in such cases, I would be fine with it.
This would of course mean that the test cases won't automatically detect if something doesn't work anymore.
Otherwise I see no problem with manual tests.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

pabulaner commented:
So are there any updates?

kevinrushforth commented:
Not yet, but it's still on my list.

kevinrushforth reviewed (commented):
I see that you made the tests manual (good), but left them as part of our automated test suite (not good). These tests will hang our automated test jobs, so cannot stay where they are.

You will need to move them somewhere underneath `tests/manual`. I recommend creating a new directory, called SystemMenu or something similar. You can either put them in the default package or create a package underneath the SystemMenu directory as you prefer.

You will either need an ant or gradle script to compile and run them, or else remove (comment out) the uses of JUnit. The latter is probably easier but do whichever you prefer.

I ran the automated tests as is manually on my macOS 26 system, and they work as expected for me. It might be helpful to have a "Fail" button in addition to "Next", but in the interest of getting this bug fix in, that could be done as a follow up.

So I think this is close to being ready to go in.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth commented:
They are in the right place now, but without build script support, they can't easily be run. We have an open enhancement to wire up manual tests to the build, but it isn't implemented yet. So each test needs to do one of two things in order to be able to run it:

1. Remove all dependencies on other packages, including (especially) Junit, and provide a main method for each of the tests, so that a developer can run "javac *.java" and then "java MyTest".
2. Provide a simple script (gradle or ant) to build and run the tests

I'm not sure which is easier. I'm out of time to finish the review before the holidays, but will pick it up next year. I'd like to see this get into jfx26 if possible.

Commit:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth reviewed (commented):
The tests all work as expected. The only thing I notice is that they don't always exit when the last window is closed, but given the nature of the tests, that isn't surprising. Improving this could be done as a follow-up.

In parallel with your enabling a GHA run, I'll do one last CI headful test run. When both are successful, I'll approve it.

andy-goryachev-oracle commented on tests/manual/systemmenu/README.md:
```diff
@@ -0,0 +1,13 @@
+# Compiling and Running
+
+To compile the tests, run the following command:
```
Actually, separate compile step is not necessary, the test apps run with the following command from within this directory (you do need to run the gradle build first):

```
java --module-path ../../../build/sdk/lib/ --add-modules javafx.controls,javafx.fxml,javafx.swing MacOSSystemMenuJFXPanelSwingFirstTest.java
```

beldenfox replied:
Even easier:

`java @../../../build/run.args MacOSSystemMenuMultiWindowTest.java `

pabulaner replied:
Thanks, very nice and clean solution!

andy-goryachev-oracle commented on tests/manual/systemmenu/MacOSSystemMenuMultiWindowWithSwingFirstTest.java:
```diff
@@ -0,0 +1,87 @@
+        initSwing(List.of(TEST_MENUS_0, TEST_MENUS_2));
+        initJavaFX(List.of(TEST_MENUS_1, TEST_MENUS_3));
+
+        focusJavaFX(0);
```
Minor: noticed that the "Next" button seem to work only once in this and other tests.  Is it intentional?

pabulaner replied:
For me this test works fine, with the next button doing exactly what it should.

andy-goryachev-oracle commented on tests/manual/systemmenu/MacOSSystemMenuSingleWindowWithSwingFirstTest.java:
```diff
@@ -0,0 +1,74 @@
+public class MacOSSystemMenuSingleWindowWithSwingFirstTest extends MacOSSystemMenuTestBase {
+
+    public static void main(String[] args) throws InterruptedException, IOException {
+        new MacOSSystemMenuSingleWindowWithSwingFirstTest().test();
```
this test hangs when fx window is closed first via the close title bar icon, followed by closing the swing window with same method.

pabulaner replied:
Again, the tests are intended only for validating the contents of the system menu bar and then pressing next. Everything else is not intended to work.

andy-goryachev-oracle reviewed (commented):
Some minor comments, one exception, and many tests hang on closing using the system menu, or the quit menu in the dock, or the close icon in the title bar.

pabulaner committed:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

pabulaner commented on tests/manual/systemmenu/MacOSSystemMenuMultiWindowTest.java:
```diff
@@ -0,0 +1,81 @@
+    public static void main(String[] args) throws InterruptedException, IOException {
+        new MacOSSystemMenuMultiWindowTest().test();
+    }
+
```
The tests are not meant for exiting via the system menu. They are intended to be run, so the user can test it, and then terminate themselves at the end.

pabulaner committed:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

pabulaner commented:
So I think I fixed all the major comments. I added a System.exit() to every test, so after pressing the next button, the windows will all close. I also edited the readme, so the user knows what is expected and how to run the tests.

pabulaner committed:
8359108: Mac - When Swing starts First, native application menu doesn't work for JavaFX

kevinrushforth reviewed (approved):
All tests now exit for me after presing "Next" the last time. The quick headful CI test I mentioned passed and I see that the GHA run has passed.

I think any further improvements to the test can be done in a follow-up issue.

beldenfox reviewed (approved):
I'll approve this as-is. I'd still like to see some cleanup and expansion of the tests (I had to test the "all windows iconified" cases separately) but that can be done later. I'll enter a bug for that sometime next week.

kevinrushforth commented:
@pabulaner This is now ready to integrate.

crschnick commented:
I just noticed that the desktop methods of AWT like `Desktop.getDesktop().setPreferencesHandler()` do no longer work in newer releases of JavaFX. When you first start AWT then JavaFX, they worked fine previously. Now, no menu option is showing up anymore. I haven't fully determined yet which build exactly introduced this regression, but this PR looks to be the most obvious one. Would that be an intentional consequence of this PR?

beldenfox commented:
@crschnick No that would not be intentional. It seems unlikely that this PR is the cause unless your app deliberately mixes top level Swing and JavaFX windows. But I can't say for sure until I have a chance to look into this next week. Do you have standalone code that reproduces the bug?

crschnick commented:
You can check out https://github.com/crschnick/jfx-test/blob/master/src/main/java/io/xpipe/jfx_runner/MenuBar.java. It uses the new jfx-runner gradle scripts, so you can run and debug it from that repo by just running ./gradlew run

beldenfox commented:
@crschnick If I understand you correctly in the past calling `Desktop.getDesktop().setPreferencesHandler()` would ensure that a Preferences item showed up in the system application menu. I can't reproduce that behavior. I've taken your MenuBar.java sample code and run it with JDK 21 and JavaFX 17 and other combinations and can't get that to happen. No matter what I do I only see the JavaFX  application menu (4 items), never the AWT one (at least 6).

BTW I'm testing your MenuBar.java outside of the jfx-runner gradle system. I've already got my own shell scripts that allow me to switch between different JFX and Java versions. I did need to tweak MenuBar.java to make `main` public.

crschnick commented:
@beldenfox Do you mean that you don't see any menu bar entry in general, or only for this specific case where the order of execution is different?

I know that the menu bar entries were correctly shown in the past in earlier JDK, JavaFX, and macOS versions. I don't have an older macOS system than 26, so I can't test it going back.

I also noticed that the Desktop setAboutHandler() method seems to be broken in general for AWT for macOS 26, so maybe this is more related to the macOS version as that method worked for earlier versions as well.

beldenfox commented:
@crschnick I never see the Settings menu item regardless of order of execution. This is true even in older versions of JavaFX, long before this PR came along.

To see the Settings menu item in a JavaFX app you need to initialize it so AWT creates the application menu instead of JavaFX. I don't see how you can do that with test code you provided even in earlier versions of JavaFX. I would like to look into this but need a stand-alone program that reproduces the problem e.g. that shows the Settings menu in an older version of JavaFX but not in the current version.

crschnick commented:
@beldenfox What macOS version are you on? I see the settings entry in the reproducer I posted on macOS 26 with JavaFX master and JDK 25. The call to `GraphicsEnvironment.getLocalGraphicsEnvironment()` will initialize AWT in the reproducer

crschnick commented:
Or more specifically, I see the settings entry in the reproducer if I uncomment the respective line where the comment says. The default state of the reproducer is to show that the settings entry is not shown, but can be when changing the order by uncommenting

beldenfox commented:
I think I understand what's going on now (it helps that I'm no longer on vacation and no longer on jury duty). Again, this PR concerns what happens when you have a top-level JavaFX window focused inside an AWT application.

In the past the only application menu you would see would be provided by AWT. The AWT application menu always has an About item and optionally has a Settings item if the client has called `Desktop.setPreferencesHandler`. With this PR when a JavaFX window is focused you'll see the JavaFX version of the application menu, not the AWT one. That might be considered a regression though it's hard to say since I think this whole setup is officially unsupported (?)

@crschnick The MenuBar.java test you provided illustrates something entirely new. When a top-level JavaFX window is focused in an AWT application and you call `Desktop.setPreferencesHandler` AWT will manipulate the JavaFX platform menu and insert the Settings item into it. The result is a weird menu with Settings in the wrong location and a mix of NSMenuItems created by Glass and AWT. Again, technically a bug but maybe the resolution is just to tell developers to not do that.

The MenuBar.java test only reproduces the issue when run inside the jfx-test harness. When run as a standalone program the issue doesn't reproduce (Settings never appears in the application menu). This is true even after I tweaked `main` to ensure it ran. In the future please provide a stand-alone reproducer.

crschnick commented:
@beldenfox Thanks for the explaination.

Would it be possible for JavaFX to not touch anything for the menu bar when no system menu bar is set from the JavaFX side? That would make these problems go away in the smoothest fashion.

The test harness shouldn't do anything differently than just launching the test program with the latest jfx changes, so I am not sure why that works differently. So it should behave identical to a standalone reproducer. I will try to figure that one out.

beldenfox commented:
> Would it be possible for JavaFX to not touch anything for the menu bar when no system menu bar is set from the JavaFX side? That would make these problems go away in the smoothest fashion.

Yes, this should be possible. I'm working on it now.
